### PR TITLE
Use slotmap instead of bespoke arena, limiting to one parent only

### LINKED ISSUE #191: Nodes can have multiple parents
Nodes can currently have multiple parents. Under the no-allocation settings though, the number of parents is limited to one, as reported by @mcrvz in #176.

This creates divergence, but the correct behavior is actually to limit the number of parents to 1 across all platforms (likely by not storing a vec). Multiple parents cannot be defined in HTML, and there is no logical interpration of how style inheritance should work with more than one parent.

This should both improve robustness and performance.

### PR DESCRIPTION
**Pull request by jkelleyrtp**
# Objective

This PR removes the ability for nodes to have multiple parents. Fixes #191

This simplifies some of the codebase, but mostly is intended to bring Taffy closer to spec for HTML.

It also removes some of the bespoke swap_remove code which seemed difficult to maintain.

The next logical step is to use hecs instead of slotmap, but I found that integrating hecs was becoming a lot of work given the code is designed mostly around a struct of arrays. Slotmap seemed to fit in better.

### CODE REVIEW

**Comment by alice-i-cecile** [comment]
Definitely curious to see before / after benchmarks for these changes :)

**Comment by jkelleyrtp** [comment]
Here's the results:
- Deep Layout Build: -60.944% 
- Deep Layout Single: -13.944%
- Deep Layout Relayout: +2.0090%
- Generated Benchmarks: -30.319%

It looks like overall a big win for building trees, but a negligible win for relayouts. Which, is what you'd expect since there was already an internal arena system. The connection between user code and internal code will be faster, at the very least.

Old:
```
deep hierarchy - build  time:   [1.8723 us 1.8736 us 1.8751 us]                                    
                        change: [-1.3236% -0.9099% -0.5534%] (p = 0.00 < 0.05)
                        Change within noise threshold.
Found 11 outliers among 100 measurements (11.00%)
  6 (6.00%) high mild
  5 (5.00%) high severe

deep hierarchy - single time:   [8.2531 us 8.2627 us 8.2743 us]                                     
                        change: [-0.6573% -0.3675% -0.0925%] (p = 0.01 < 0.05)
                        Change within noise threshold.
Found 13 outliers among 100 measurements (13.00%)
  1 (1.00%) low mild
  6 (6.00%) high mild
  6 (6.00%) high severe

deep hierarchy - relayout                                                                             
                        time:   [4.2865 us 4.2923 us 4.2985 us]
                        change: [-0.7002% -0.4260% -0.1599%] (p = 0.00 < 0.05)
                        Change within noise threshold.
Found 6 outliers among 100 measurements (6.00%)
  2 (2.00%) high mild
  4 (4.00%) high severe

generated benchmarks    
                        time:   [304.68 us 305.43 us 306.18 us]                                 
                        change: [+0.0310% +0.4128% +0.7712%] (p = 0.03 < 0.05)
                        Change within noise threshold.
Found 2 outliers among 100 measurements (2.00%)
  2 (2.00%) high mild
```

New:
```
deep hierarchy - build  time:   [730.92 ns 732.54 ns 734.13 ns]                                    
                        change: [-61.059% -60.944% -60.815%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 6 outliers among 100 measurements (6.00%)
  6 (6.00%) high mild

deep hierarchy - single time:   [7.1005 us 7.1179 us 7.1371 us]                                     
                        change: [-14.192% -13.944% -13.690%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 14 outliers among 100 measurements (14.00%)
  10 (10.00%) high mild
  4 (4.00%) high severe

deep hierarchy - relayout                                                                             
                        time:   [4.3750 us 4.3803 us 4.3858 us]
                        change: [+1.7462% +2.0090% +2.2782%] (p = 0.00 < 0.05)
                        Performance has regressed.
Found 9 outliers among 100 measurements (9.00%)
  1 (1.00%) high mild
  8 (8.00%) high severe

generated benchmarks    
                        time:   [212.28 us 212.66 us 213.05 us]                                 
                        change: [-30.531% -30.319% -30.090%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 6 outliers among 100 measurements (6.00%)
  3 (3.00%) high mild
  3 (3.00%) high severe
```

**Comment by jkelleyrtp** [comment]
@alice-i-cecile Should we use stable keys or unstable keys for the tree? The current implementation uses stable, I believe. This means that a key for a node that was removed because of a parent removal will throw a hard error when used later. Unstable keys mean you could be accidentally manipulating new nodes using old keys if you weren't careful.

We could either adopt a generational arena, which I think slows down performance a tad, or just clearly document that Taffy uses a slab and that it is incorrect usage for outdated keys to be used once removed.

**Comment by alice-i-cecile** [comment]
I would prefer to use stable keys for now as the conservative choice. Once that's in place we can experiment with swapping to unstable keys, and see if the benchmarks justify it.

WRT error handling, I would like to try and catch + propagate errors as aggressively as we can for now. This code base is a pain to debug and maintain, and moving away from panicking index operations will make fixing bugs much easier. Furthermore panics in the layout often means that an entire game or GUI is crashing, which is not an acceptable failure mode for a layouting library because of how inconsequential mistakes here are compared to things like app state.

File: src/error.rs [hunk 1]
```diff
@@ -4,23 +4,12 @@ use core::fmt::{Display, Formatter, Result};
...
-impl Display for InvalidNode {
-    fn fmt(&self, f: &mut Formatter) -> Result {
-        write!(f, "Node {:?} is not in the Taffy instance", self.0)
-    }
-}
-
-#[cfg(feature = "std")]
-impl std::error::Error for InvalidNode {}
+/// The error Taffy generates on invalid operations
+pub type TaffyResult<T> = core::result::Result<T, TaffyError>;
```
Comment (alice-i-cecile) [inline src/error.rs:8]: I don't think this type alias makes things clearer.

File: src/error.rs [hunk 2]
```diff
@@ -4,23 +4,12 @@ use core::fmt::{Display, Formatter, Result};
...
 
 /// An error that occurs while trying to access or modify a [`Node`]'s children by index.
```
Comment (alice-i-cecile) [inline src/error.rs:10]: These docs need updating.

File: src/error.rs [hunk 3]
```diff
@@ -4,23 +4,12 @@ use core::fmt::{Display, Formatter, Result};
...
 #[derive(Debug)]
-pub enum InvalidChild {
+pub enum TaffyError {
```
Comment (alice-i-cecile) [inline src/error.rs:12]: This change needs to get noted in the release notes.

File: src/flexbox.rs [hunk 4]
```diff
@@ -1,22 +1,25 @@
-//! Computes the [flexbox](https://css-tricks.com/snippets/css/a-guide-to-flexbox/) layout algorithm on a [`Forest`](crate::forest::Forest) according to the [spec](https://www.w3.org/TR/css-flexbox-1/)
+//! Computes the [flexbox](https://css-tricks.com/snippets/css/a-guide-to-flexbox/) layout algorithm on a [`Forest`](crate::Taffy::Forest) according to the [spec](https://www.w3.org/TR/css-flexbox-1/)
```
Comment (alice-i-cecile) [inline src/flexbox.rs:1]: This doc link doesn't look right.

**Review by alice-i-cecile — CHANGES_REQUESTED** [review]
Awesome! General requests:

1. Can you add `#[forbid(unsafe_code)]` to the crate now that indexmap is gone?
2. I'd like to see some benchmarks, but won't block on it. I don't care if this regresses performance given the removal of unsafe + code quality improvements, but it's good marketing.
3. I like the use of `Node` much more than `NodeId`.
4. Getting rid of all that code in `Forest` feels great.
5. Only one parent is a nice correctness win.

**Comment by jkelleyrtp** [comment]
> Awesome! General thoughts:
> 
> 1. Can you add `#[forbid(unsafe_code)]` to the crate now that indexmap is gone?
> 2. I'd like to see some benchmarks, but won't block on it. I don't care if this regresses performance given the removal of unsafe + code quality improvements, but it's good marketing.
> 3. I like the use of `Node` much more than `NodeId`.
> 4. Getting rid of all that code in `Forest` feels great.
> 5. Only one parent is a nice correctness win.

Sure - I also need to fix some of the documentation surrounding "forests" as you've noted.

Per 2 - I posted above https://github.com/DioxusLabs/taffy/pull/196#issuecomment-1171868429. Roughly a 30-90% win from average to best case. Worst cases hover around 0-15%. I tried out slab but went to slotmap and saw no performance difference, and Slotmap keys are the same sizes as slab keys. They also benefit from nonzero optimizations, so the `Option<Node>` for parents is the same size as before. Essentially we get stable keys for free.

File: RELEASES.md [hunk 5]
```diff
@@ -20,6 +20,11 @@
 - renamed `taffy::style::Style` -> `taffy::style::FlexboxLayout` to more precicely indicate its purpose
 - renamed `taffy::Error` -> `taffy::error::InvalidNode`
```
Comment (alice-i-cecile) [inline RELEASES.md:21]: This needs to be renamed here.

File: RELEASES.md [hunk 6]
```diff
@@ -20,6 +20,11 @@
...
 - `taffy::Taffy::remove_child_at_index`, `taffy::Taffy::replace_child_at_index`, and `taffy::Taffy::child_at_index` now return `taffy::InvalidChild::ChildIndexOutOfBounds` instead of panicing
+- `taffy::Node` is now unique only to the Taffy instance from which it was created.
+- `taffy::error::InvalidChild` is now `taffy::error::TaffyError`
+- `taffy::error::InvalidNode` has been removed and is now just a branch on the `TaffyError` enum
+- `taffy::forest::Forest` has been merged into `taffy::node::Taffy` for a performance boost up to 90%
+
 
 ### 0.2.0 Fixed
```
Comment (alice-i-cecile) [inline RELEASES.md:29]: I would add a note about only one parent here.

File: RELEASES.md (hunk 5 above)
Comment (alice-i-cecile) [reply RELEASES.md:21]:
```suggestion
- renamed `taffy::Error` -> `taffy::error::TaffyError`
```

File: RELEASES.md (hunk 6 above)
Comment (alice-i-cecile) [reply RELEASES.md:29]:
```suggestion
### 0.2.0 Fixed

- nodes can only ever have one parent
```

**Review by alice-i-cecile — APPROVED** [review]
This is good to go!

File: src/node.rs [hunk 7]
```diff
@@ -87,255 +53,232 @@ impl Default for Taffy {
...
-    fn find_node(&self, node: Node) -> Result<NodeId, error::InvalidNode> {
-        match self.nodes_to_ids.get(&node) {
-            Some(id) => Ok(*id),
-            None => Err(error::InvalidNode(node)),
-        }
-    }
+    /// Creates and adds a new unattached leaf node to the tree, and returns the [`NodeId`] of the new node
+    pub fn new_leaf(&mut self, layout: FlexboxLayout) -> TaffyResult<Node> {
+        let id = self.nodes.insert(NodeData::new(layout));
+        let _ = self.children.insert(new_vec_with_capacity(0));
```
Comment (Sheepyhead) [inline src/node.rs:79]: What is the reason for the dropped binding here?

File: src/node.rs [hunk 8]
```diff
@@ -87,255 +53,232 @@ impl Default for Taffy {
...
     }
 
     /// Removes all nodes
     ///
     /// All associated [`Id`] will be rendered invalid.
     pub fn clear(&mut self) {
-        self.nodes_to_ids.clear();
-        self.ids_to_nodes.clear();
-        self.forest.clear();
+        self.nodes.clear();
```
Comment (Sheepyhead) [inline src/node.rs:114]: Much cleaner, much more readable, lovely!

File: src/node.rs [hunk 9]
```diff
@@ -87,255 +53,232 @@ impl Default for Taffy {
...
     }
 
     /// Marks the layout computation of this node and its children as outdated
-    pub fn mark_dirty(&mut self, node: Node) -> Result<(), error::InvalidNode> {
-        let id = self.find_node(node)?;
-        self.forest.mark_dirty(id);
+    pub fn mark_dirty(&mut self, node: Node) -> TaffyResult<()> {
+        /// Performs a recursive depth-first search up the tree until the root node is reached
+        ///
+        ///  WARNING: this will stack-overflow if the tree contains a cycle
```
Comment (Sheepyhead) [inline src/node.rs:256]: Should this comment be pulled up into the function comment so it's visible for users of the API?

File: RELEASES.md [hunk 10]
```diff
@@ -18,11 +18,18 @@
...
 
+- nodes can only ever have one parent
+
```
Comment (alice-i-cecile) [inline RELEASES.md:32]:
```suggestion
```

File: RELEASES.md [hunk 11]
```diff
@@ -18,11 +18,18 @@
...
 - renamed `taffy::node::Taffy.new-node(..)` -> `taffy::node::Taffy.new_with_children(..)`
 - renamed `taffy::style::Style` -> `taffy::style::FlexboxLayout` to more precicely indicate its purpose
-- renamed `taffy::Error` -> `taffy::error::InvalidNode`
+- renamed `taffy::Error` -> `taffy::error::TaffyError`
 - `taffy::Taffy::remove_child_at_index`, `taffy::Taffy::replace_child_at_index`, and `taffy::Taffy::child_at_index` now return `taffy::InvalidChild::ChildIndexOutOfBounds` instead of panicing
+- `taffy::Node` is now unique only to the Taffy instance from which it was created.
+- `taffy::error::InvalidChild` is now `taffy::error::TaffyError`
+- `taffy::error::InvalidNode` has been removed and is now just a branch on the `TaffyError` enum
+- `taffy::forest::Forest` has been merged into `taffy::node::Taffy` for a performance boost up to 90%
+
```
Comment (alice-i-cecile) [inline RELEASES.md:27]:
```suggestion
```

File: src/node.rs (hunk 7 above)
Comment (jkelleyrtp) [reply src/node.rs:79]: The key for each slotmap ends up the same but .insert is labeled must_use so either we do a drop binding or we call drop manually.

I turned on clippy pedantic just to poke around and it suggested an explicit drop... in any case all of the keys are the same.

**Review by Sheepyhead — APPROVED** [review]

**Pull request merged**